File: client-ios/SerenadaCore/Sources/Call/AudioLevelPoller.swift
import Combine
import Foundation

/// Polls the primer peer connection for local `media-source.audioLevel`
/// (via `collectLocalLevel`) and each peer slot for inbound (remote) level
/// every `AudioLevelMonitor.updateIntervalSeconds`, runs the raw values
/// through per-cid `AudioLevelMonitor`s for dBFS+EMA smoothing, and reports
/// the smoothed values on the main actor.
///
/// The primer keeps WebRTC's audio pipeline live throughout a session, so
/// the local level is consistent across Waiting and InCall — and works on
/// the very first tick after a peer leaves and we're back to alone in the
/// room. Mirrors the web SDK's `AudioLevelMonitor` smoothing pipeline so
/// the indicator visual is consistent across platforms.
@MainActor
final class AudioLevelPoller {
    private let clock: SessionClock
    private let isActivePhase: () -> Bool
    private let getPeerSlots: () -> [any PeerConnectionSlotProtocol]
    private let collectLocalLevel: (@escaping @Sendable (Float?) -> Void) -> Void
    private let onLevelsUpdated: (_ localLevel: Float, _ remoteLevels: [String: Float]) -> Void

    private var pollTimerCancellable: AnyCancellable?
    private var requestInFlight = false
    private let localMonitor = AudioLevelMonitor()
    private var remoteMonitors: [String: AudioLevelMonitor] = [:]
    /// Bumped on each `start()`/`stop()` so any stats request started in a
    /// previous run can be discarded when its async result returns.
    private var generation: UInt64 = 0

    init(
        clock: SessionClock,
        isActivePhase: @escaping () -> Bool,
        getPeerSlots: @escaping () -> [any PeerConnectionSlotProtocol],
        collectLocalLevel: @escaping (@escaping @Sendable (Float?) -> Void) -> Void,
        onLevelsUpdated: @escaping (_ localLevel: Float, _ remoteLevels: [String: Float]) -> Void
    ) {
        self.clock = clock
        self.isActivePhase = isActivePhase
        self.getPeerSlots = getPeerSlots
        self.collectLocalLevel = collectLocalLevel
        self.onLevelsUpdated = onLevelsUpdated
    }

    func start() {
        stop()
        generation &+= 1
        pollTimerCancellable = clock.scheduleRepeating(intervalSeconds: AudioLevelMonitor.updateIntervalSeconds) { [weak self] in
            self?.tick()
        }
    }

    func stop() {
        pollTimerCancellable?.cancel()
        pollTimerCancellable = nil
        // Bumping the generation invalidates any stats request kicked off by
        // a previous tick — when its async completion fires, applyAndEmit
        // will drop the result instead of emitting after teardown.
        generation &+= 1
        requestInFlight = false
        localMonitor.reset()
        remoteMonitors.removeAll()
    }

    private func tick() {
        guard pollTimerCancellable != nil, isActivePhase() else { return }
        if requestInFlight { return }
        let slots = getPeerSlots()

        requestInFlight = true
        let tickGeneration = generation
        let group = DispatchGroup()
        let lock = NSLock()
        var rawRemote: [String: Float?] = [:]
        var rawLocal: Float?

        group.enter()
        collectLocalLevel { level in
            lock.lock()
            rawLocal = level
            lock.unlock()
            group.leave()
        }

        for slot in slots {
            let cid = slot.remoteCid
            group.enter()
            slot.collectAudioLevels { inbound, _ in
                lock.lock()
                rawRemote[cid] = inbound
                lock.unlock()
                group.leave()
            }
        }

        group.notify(queue: .main) { [weak self] in
            guard let self else { return }
            Task { @MainActor in self.applyAndEmit(rawLocal: rawLocal, rawRemote: rawRemote, tickGeneration: tickGeneration) }
        }
    }

    private func applyAndEmit(rawLocal: Float?, rawRemote: [String: Float?], tickGeneration: UInt64) {
        // Drop late results from a previous start()/stop() cycle.
        guard tickGeneration == generation, pollTimerCancellable != nil else {
            requestInFlight = false
            return
        }
        requestInFlight = false
        let local = localMonitor.update(rawLevel: rawLocal ?? 0)
        // Drop monitors for peers no longer present. Snapshot the keys so we
        // don't mutate the dictionary while iterating it.
        let activeCids = Set(rawRemote.keys)
        for cid in Array(remoteMonitors.keys) where !activeCids.contains(cid) {
            remoteMonitors.removeValue(forKey: cid)
        }
        var remote: [String: Float] = [:]
        for (cid, raw) in rawRemote {
            let monitor = remoteMonitors[cid] ?? AudioLevelMonitor()
            remoteMonitors[cid] = monitor
            remote[cid] = monitor.update(rawLevel: raw ?? 0)
        }
        onLevelsUpdated(local, remote)
    }
}

File: client-android/serenada-core/src/main/java/app/serenada/core/call/AudioLevelPoller.kt
package app.serenada.core.call

import android.os.Handler
import java.util.concurrent.ExecutorService

/**
 * Polls WebRTC stats every [AudioLevelMonitor.UPDATE_INTERVAL_MS] for the
 * local media-source level (via [collectLocalAudioLevel]) and each remote
 * peer's inbound level (via the slot's `collectAudioLevels`), pushes the
 * raw values through per-cid [AudioLevelMonitor]s for dBFS+EMA smoothing,
 * and reports the results on the main handler.
 *
 * The local level is sourced from a primer peer connection that stays alive
 * the entire time the user is in a room, so `media-source.audioLevel`
 * reads consistently — including the Waiting phase before any real peer
 * joins. Mirrors the web SDK's `AudioLevelMonitor` smoothing pipeline so
 * the indicator visual is consistent across platforms.
 */
internal class AudioLevelPoller(
    private val handler: Handler,
    private val statsExecutorProvider: () -> ExecutorService?,
    private val isActivePhase: () -> Boolean,
    private val getPeerSlots: () -> List<PeerConnectionSlotProtocol>,
    private val collectLocalLevel: ((Float?) -> Unit) -> Unit,
    private val onLevelsUpdated: (localLevel: Float, remoteLevels: Map<String, Float>) -> Unit,
) {
    private val localMonitor = AudioLevelMonitor()
    private val remoteMonitors = mutableMapOf<String, AudioLevelMonitor>()
    private var pollRunnable: Runnable? = null
    private var requestInFlight = false

    fun start() {
        if (pollRunnable != null) return
        val runnable = object : Runnable {
            override fun run() {
                tick()
                handler.postDelayed(this, AudioLevelMonitor.UPDATE_INTERVAL_MS)
            }
        }
        pollRunnable = runnable
        handler.post(runnable)
    }

    fun stop() {
        pollRunnable?.let { handler.removeCallbacks(it) }
        pollRunnable = null
        requestInFlight = false
        localMonitor.reset()
        remoteMonitors.clear()
    }

    private fun tick() {
        if (!isActivePhase()) return
        if (requestInFlight) return
        val slots = getPeerSlots()
        val executor = statsExecutorProvider()?.takeIf { !it.isShutdown }
        if (executor == null) {
            // Executor torn down mid-session: emit a fully decayed sample so
            // the indicator drops to silence rather than freezing.
            val local = localMonitor.update(0f)
            val remote = remoteMonitors.mapValues { (_, m) -> m.update(0f) }
            onLevelsUpdated(local, remote)
            return
        }
        requestInFlight = true
        try {
            executor.execute { collectAndReport(slots) }
        } catch (_: java.util.concurrent.RejectedExecutionException) {
            requestInFlight = false
        }
    }

    private fun collectAndReport(slots: List<PeerConnectionSlotProtocol>) {
        val rawRemote = mutableMapOf<String, Float?>()
        var rawLocal: Float? = null
        var remaining = slots.size + 1  // +1 for the primer query
        val sync = Any()

        val finishOne = {
            synchronized(sync) {
                remaining -= 1
                if (remaining == 0) {
                    val capturedLocal = rawLocal
                    val capturedRemote = rawRemote.toMap()
                    handler.post { applyAndEmit(capturedLocal, capturedRemote) }
                }
            }
        }

        collectLocalLevel { level ->
            synchronized(sync) { rawLocal = level }
            finishOne()
        }

        slots.forEach { slot ->
            slot.collectAudioLevels { inbound, _ ->
                synchronized(sync) { rawRemote[slot.remoteCid] = inbound }
                finishOne()
            }
        }
    }

    private fun applyAndEmit(rawLocal: Float?, rawRemote: Map<String, Float?>) {
        requestInFlight = false
        val local = localMonitor.update(rawLocal ?: 0f)
        // Drop monitors for peers no longer present.
        val activeCids = rawRemote.keys
        remoteMonitors.keys.retainAll(activeCids)
        val remote = rawRemote.mapValues { (cid, raw) ->
            val monitor = remoteMonitors.getOrPut(cid) { AudioLevelMonitor() }
            monitor.update(raw ?: 0f)
        }
        onLevelsUpdated(local, remote)
    }
}
